cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/artifact/ArtifactRepository.java

```diff
@@ -349,13 +353,92 @@ public ArtifactDetail addArtifact(Id.Artifact artifactId, File artifactFile,
 
     try {
       ArtifactClasses artifactClasses = inspectArtifact(artifactId, artifactFile, additionalPlugins, parentClassLoader);
-      ArtifactMeta meta = new ArtifactMeta(artifactClasses, parentArtifacts);
+      Map<String, String> artifactProperties = properties == null ? Collections.<String, String>emptyMap() : properties;
+      ArtifactMeta meta = new ArtifactMeta(artifactClasses, parentArtifacts, artifactProperties);
       return artifactStore.write(artifactId, meta, Files.newInputStreamSupplier(artifactFile));
     } finally {
       parentClassLoader.close();
     }
   }
 
+  /**
+   * Writes properties of an artifact. Any existing properties will be overwritten
```

Comment:
In the error messages above, you used "for an artifact" and here and below you use "of an artifact". I think we should be consistent, no? Which better describes them? Are the properties separate from the artifact, or are they stored with them? It seems that the latter is the case from what I read here...